+sw_tests/+unit_tests/unittest_spinw_genmagstr.m

```diff
@@ -0,0 +1,519 @@
+        end
+        function test_direct_multiatom_nExt(testCase)
+            swobj = copy(testCase.swobj);
+            swobj.addatom('r', [0.5 0.5 0.5], 'S', 2);
+            S = [1  0  1 -1; ...
+                 1  1  0  0; ...
+                 0 -1  0  0];
+            nExt = [int32(2) int32(1) int32(1)];
+            k = [0 1/3 0];
+            swobj.genmagstr('mode', 'direct', 'S', S, 'nExt', nExt, 'k', k);
+            expected_mag_str = struct('nExt', nExt, ...
+                                      'k', k', ...
+                                      'F', [sqrt(2)/2        0 1 -2; ...
+                                            sqrt(2)/2  sqrt(2) 0  0; ...
+                                                    0 -sqrt(2) 0  0]);
+            testCase.verify_obj(expected_mag_str, swobj.mag_str);
+        end
+        function test_direct_multiatom_multik(testCase)
+            swobj = copy(testCase.swobj);
+            swobj.addatom('r', [0.5 0.5 0.5], 'S', 2);
+            S_k = [1 0; 1 1; 0 -1];
+            S = cat(3, S_k, S_k);
+            k = [0 1/3 0; 1/2 0 0];
+            swobj.genmagstr('mode', 'direct', 'S', S, 'k', k);
+            F_k = [sqrt(2)/2        0; ...
+                   sqrt(2)/2  sqrt(2); ...
+                           0 -sqrt(2)];
+            expected_mag_str = testCase.default_mag_str;
+            expected_mag_str.k = k';
+            expected_mag_str.F = cat(3, F_k, F_k);
+            testCase.verify_obj(expected_mag_str, swobj.mag_str);
+        end
+        function test_direct_multik_scalar_nExt(testCase)
+            % Test if a scalar is used for nExt it is treated as a
+            % tolerance to automatically determine nExt
+            swobj = copy(testCase.swobj);
+            S_k = [1 0 1 0 1 0; 0 0 1 1 0 0; 0 0 0 1 1 1];
+            S = cat(3, S_k, S_k);
+            nExt = 0.01;
+            k = [0 1/3+0.5*nExt 0; 1/2+0.5*nExt 0 0];
+            swobj.genmagstr('mode', 'direct', 'S', S, 'k', k, 'nExt', nExt);
+            F_k = [1 0 sqrt(2)/2         0 sqrt(2)/2 0; ...
+                   0 0 sqrt(2)/2 sqrt(2)/2         0 0; ...
+                   0 0         0 sqrt(2)/2 sqrt(2)/2 1];
+            expected_mag_str = struct('nExt', [int32(2) int32(3) int32(1)], ...
+                                      'k', k', ...
+                                      'F', cat(3, F_k, F_k));
+            testCase.verify_obj(expected_mag_str, swobj.mag_str);
+        end
+        function test_helical_tri(testCase)
+            swobj_tri = copy(testCase.swobj_tri);
+            swobj_tri.genmagstr('mode', 'helical', 'S', [1; 0; 0], ...
+                                'k', [1/3 1/3 0]);
+            expected_mag_str = testCase.default_mag_str;
+            expected_mag_str.k = [1/3; 1/3; 0];
+            expected_mag_str.F = [1.5; 1.5i; 0];
+            testCase.verify_obj(expected_mag_str, swobj_tri.mag_str);
+        end
+        function test_helical_tri_lu_unit(testCase)
```

Comment:
I have a `.mcif` for CeAuSb2 - which has q=[0.135,0.135,0.5] and moment along c  - would that work to show the difference between lattice and cartesian frames? I also have a structure with q=[0.25,0,0.5] (moment along c again). I'm guessing commensurate is better?

Reply:
Thanks, does the commensurate structure have a non-Cartesian lattice (monoclinic or lower)? It's just that for orthrohombic or tetragonal (or hexagonal with moments along c) the moment directions in lattice units and in SpinW's crystal Cartesian will be the same.

Reply:
Ah I see, no CeAuSb2 is cartesian in that respect, but I think I have a .mcif for PdCrO2 somewhere which has a triangular lattice and quite a complicated structure [https://iopscience.iop.org/article/10.1088/1367-2630/ac4280/meta]
![image](https://user-images.githubusercontent.com/55979119/183604646-bc552aca-b404-43a2-a64f-82ee683618fc.png)

Reply:
Yes, that would be great!

I also actually have a SpinW model for that system, although not with the canting...

Reply:
Ah if you already have a spinw model then we can use that! Might be easier to see what's going on without the canting tbh, but I'll have a look for the mcif anyway.

Reply:
@mducle @RichardWaiteSTFC Do we have the mcif/spinw model for this system?

Reply:
Ah... so Richard will get an `mcif` file but we don't currently have an `mcif` reader, so I propose to create a new issue (feature request) to read in mcif files and to double check the handling of `lu` units in that instead of this PR.

Reply:
https://github.com/SpinW/spinw/issues/104